app/src/main/java/com/nio/TcpDataFlowExample.java

```diff
@@ -1,117 +1,105 @@
-						final SocketChannel client = server.accept();
-						client.configureBlocking(false);
-						System.out.printf("New connection accepted: %s%n", client);
-
-						clients.put(server.socket().getLocalPort(), client);
-
-						client.register(selector, SelectionKey.OP_READ);
-					}
-
-					if (selectionKey.isReadable()) {
-						System.out.println("Handle READ event");
-						final SocketChannel client = (SocketChannel) selectionKey.channel();
-						final int read = client.read(buffer);
-
-						if (read == -1) {
-							client.close();
-							client.keyFor(selector).cancel();
-							System.out.printf("The connection was closed: %s%n", client);
-							return;
-						}
-
-						buffer.flip();
-						switch (client.socket().getLocalPort()) {
-						case 5555:
-							for (int i = 0; i < buffer.limit(); i++) {
-								buffer.put(i, (byte) Character.toUpperCase(buffer.get(i)));
-							}
-							client.write(buffer);
-							break;
-						case 4444:
-							final String command = StandardCharsets.UTF_8.decode(buffer).toString();
-							final String stopCommand = "stop-read";
-							if (stopCommand.equals(command.trim().toLowerCase())) {
-								System.out.println("Handle stop-read");
-								registerEvent(selector, clients, 0);
-								break;
-							}
-							final String startCommand = "start-read";
-							if (startCommand.equals(command.trim().toLowerCase())) {
-								System.out.println("Handle start-read");
-								registerEvent(selector, clients, SelectionKey.OP_READ);
-								break;
-							}
-
-							final byte[] unknownCommand = String.format("Supported commands:%n%s%n%s%n", stopCommand, startCommand)
-									.getBytes(StandardCharsets.UTF_8);
-							client.write(ByteBuffer.wrap(unknownCommand));
-							break;
-						}
-
-						buffer.clear();
-					}
-					iterator.remove();
-				} catch (Exception e) {
-					e.printStackTrace();
-				}
-			});
-		}
-	}
-
-	private static void registerEvent(final Selector selector, final Multimap<Integer, SocketChannel> clients, final int op) {
-		clients.get(5555).forEach(c -> {
-			try {
-				c.register(selector, op);
-			} catch (Exception e) {
-				e.printStackTrace();
-			}
-		});
-	}
+    public static final String HOSTNAME = "0.0.0.0";
+    private static final int byteBufferCapacity = 65535;
+    private final String hostName;
+    private final int controlPort;
+    private final int dataPort;
+    private final int[] ports;
+    private Selector selector;
+    private ByteBuffer buffer;
+    private ClientHelper clientHelper;
+
+    private TcpDataFlowExample(String hostName, int controlPort, int dataPort) {
+        this.hostName = hostName;
+        this.controlPort = controlPort;
+        this.dataPort = dataPort;
+        this.ports = new int[]{controlPort, dataPort};
+    }
+
+    public static void main(final String... args) throws Exception {
+        new TcpDataFlowExample(HOSTNAME, 5555, 4444).startEvents();
```

Comment:
Looks good! Only one remark, in practice for the servers using method bind and for the clients connect instead startEvents...